### ISSUE TITLE
Some video files fail to load at 720p, with 503 error

### ISSUE THREAD

**Issue by user234683** [user234683/youtube-local#40]
Sometimes googlevideo.com/videoplayback URLs fail with 503 error code.

- This only happens on a subset of videos, but for these it's reproducible for several days (or longer).
- I've encountered the issue at 720p, but haven't seen it happen at 360p.

I looked into it a bit, and found that different browsers report the failure differently. Waterfox doesn't give any errors in the network console, just says the video format isn't supported. Firefox displays a 503 on one of the requests. Ungoogled chromium puts (failed) in place of the status, with a tooltip of "net::ERR_CONTENT_LENGTH_MISMATCH".

For a sample video with the problem, when I wget the URL, it gives "Read error at byte 111496/138135408 (The TLS connection was non-properly terminated.)" and then retries. I doubt the server is actually returning a 503 code, I think the browser/tools are just putting that there as a placeholder. With wget, I noticed for that video that it always fails twice at the same byte positions: 111496 and 372692, which are very early into the file, even with different exit nodes or not using tor. Inspecting these positions in a hex editor, there are repeating units of 00 00 01 ** where ** is some arbitrary byte. Further into the video file is presumably the video data which is much more random looking. I'll have to investigate the mp4 format more deeply. Since this is a TLS error, it will also be worth taking a look at what happens in Wireshark.

As a temporary workaround, I'll try getting video and audio stream muxing with MediaSource working when I get some time, since the error doesn't seem to happen with the video-only urls.

**Comment by zrose584** [comment]
Shouldn't [HTTP range requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests) be indifferent to the paylouad? So why would the file structure (in this case mp4) matter?
To me it looks like a server side problem.. Probably its not fixed because their (official) clients use separate sources, so nobody notices (+ only a subset of videos is affected).

**Comment by user234683** [comment]
That's what I would expect, but perhaps the CDN servers optimize how the video file is returned based on the structure of the file (after all, why should the files with combined audio+video have this problem but the others don't, etc.). It's one avenue worth exploring.

**Comment by zrose584** [comment]
But remember that since we will (probably) never see their backend's source-code, all we can do is make guesses...
Like for example that their backend engine has problems with [H.264 Start Code Emulation](https://www.ramugedia.com/start-code-emulation).
Have you figured out the discrepancy in content length yet?

**Comment by user234683** [comment]
Update: I encountered a sample video with the problem ( https://www.youtube.com/watch?v=GS7YvR4t-UU ) and recorded what happens with Wireshark. The client sends a GET request for the video file. Notably, it sends it with a keep alive header. The server response headers are as follows:
```HTTP/1.1 200 OK
Last-Modified: Sat, 12 May 2018 16:01:39 GMT
Content-Type: video/mp4
Date: Fri, 18 Dec 2020 01:30:30 GMT
Expires: Fri, 18 Dec 2020 01:30:30 GMT
Cache-Control: private, max-age=14719
Accept-Ranges: bytes
Content-Length: 4333558
Connection: close
Alt-Svc: h3-29=":443"; ma=2592000,h3-T051=":443"; ma=2592000,h3-Q050=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"
Vary: Origin
Cross-Origin-Resource-Policy: cross-origin
X-Content-Type-Options: nosniff
Server: gvs 1.0

```
The only interesting thing here is the Connection: close header. This just means the server will close the connection after the response is finished sending (so it's ignoring the keepalive). However, unlike the previous video with this problem, where it stopped sending content early in the video file, in this case the server sends no content at all after those headers. It just sends those headers in a TCP packet, and then proceeds to send a TCP FIN to close the connection. Contrary to my expectations, there wasn't a TCP error happening here. It seems more like there's an error retrieving the content in the CDN or something so it just sends no content (a malformed HTTP response, but registered as a TCP error by most clients since the server could have sent additional TCP packets with the content but chose to FIN the connection).

The TCP close sequence doesn't match what I initially expected based on my reading but it's unlikely to be related to the problem. Here's my notes:

---

After receiving the FIN, the client then sends a TLS close notify. Then it sends an acknowledgement of the server's FIN and sends its own FIN. Unusually, the server doesn't respond with a final ACK to the client's FIN. Instead it sends two RST (reset connection) TCP packets. [This page](https://www.pico.net/kb/what-is-a-tcp-reset-rst) says RSTs are sent in response to an unexpected packet, which could be because "The packet arrives on a TCP connection that was previously established, but the local application already closed its socket or exited and the OS closed the socket." I assume this is because the server closed the socket prematurely before the TLS close notify packet was received. [This seems to be common behavior](https://security.stackexchange.com/a/82044) of ignoring the TLS RFC spec on close notify I think, so there's unlikely anything wrong here. The RFC for TCP (not to be confused with TLS) calls this a half-duplex close sequence:
```
            A host MAY implement a "half-duplex" TCP close sequence, so
            that an application that has called CLOSE cannot continue to
            read data from the connection.  If such a host issues a
            CLOSE call while received data is still pending in TCP, or
            if new data is received after CLOSE is called, its TCP
            SHOULD send a RST to show that data was lost.
```
Essentially it's signaling that the data (the close-notify data) was ignored. I think this is [why](https://stackoverflow.com/questions/32113065/gnutls-error-110-the-tls-connection-was-non-properly-terminated#32114139) wget says "the TLS connection was improperly terminated".

---

In that sample video, if you open the downloads dropdown, a lot of the video-only formats have no file size reported, which is unusual. Usually, only the 720p video+audio format has that problem. All the formats without a reported file size have this zero-content issue, so this is not only an issue with the integrated 720p format.

Next, if you inspect the playerResponse json, you'll see that it has a dashManifestUrl in streamingData, so what's likely happening is that some videos are stored the same way as live streams for some reason, and it takes time for googlevideo to process them into non-live formats, or they're never processed. Hence why the issue goes away eventually for ~50% of videos that have the problem. There is code in youtube-local in extract_info already that detects this issue for videos that were recently livestreamed, and gives the livestream URLs instead. The code checks if the video was live previously (isLiveContent key in playerResponse) and if *all* the URLs are missing content lengths. Neither conditions are true in the two example videos. Not sure why they insist on DASH even though they weren't livestreamed.

Successive retries don't work on this example video, so it looks like the solution is going to be to implement DASH.

This issue may or may not be related to a similar problem with video files (the 360p ones) not loading for some videos uploaded around 2010. The video file still works in VLC though. I'll see what happens for those once I find an example again. Also, if anyone finds more example URLs with this class of problems post them here so I can confirm my theories.

**Comment by user234683** [comment]
Update2: Just after I posted that, zrose's original example url started having the issue again and I was able to get a Wireshark capture of it. Same idea except it sends only some of the video content. With my connection the video came in TCP chunks with data length of 1208 bytes, but then the last one is smaller at 122 bytes, suggesting that there were some socket.write() calls or something and that they decided to stop at exactly that position in the video file, rather than say cancelling when more 1208 long TCP chunks were planned. wget shows it (still) failing at those same positions 111496 and 372692 even after a month. Also, despite all of this, there is no DASH url for this video so it can't just be that. And all the other formats work and have their content lengths. wget manages to get around this truncated content problem by issuing a range request for the new missing content it wants (e.g. range request for 111496 and onward). I should be able to implement that in the video routing code.
